website/docs/r/signalr_service.html.markdown

```diff
@@ -73,6 +82,18 @@ A `features` block supports the following:
 
 ---
 
+An `upstream_setting` block supports the following:
+
+* `url_template` - (Required) The upstream URL Template. Possible values are `EnableConnectivityLogs`, `EnableMessagingLogs`, and `ServiceMode`.
+
+* `category_pattern` - (Optional) The matching pattern for category names. Possible values are `*`, multiple categories separated by commas `,`, and single categories. Ex: `connections,messages`, `*`, and `connections`.
+
+* `event_pattern` - (Optional) The matching pattern for event names.  Possible values are `*`, multiple events separated by commas `,`, and single events. Ex: `connect,disconnect`, `*`, and `connect`.
+
+* `hub_pattern` - (Optional) The matching mattern for hub names.  Possible values are `*`, multiple hubs separated by commas `,`, and single hubs. Ex: `hub1,hub1`, `*`, `hub1`.
```

Comment:
pattern doesn't really make sense as its really just lists, so could we make these lists instead of strings and join them in code?
```suggestion
* `categories` - (Optional) The categories to match on, or `*` for all.

* `events` - (Optional) The events to match on, or `*` for all.

* `hubs` - (Optional) The hubs to match on, or `*` for all.
```

and then use would be
```
 upstream_setting {
    category_pattern = ["connections", "messages"]
    event_pattern    = ["*"]
    hub_pattern      = ["hub1"]
    url_template     = "http://foo.com" // if this is just any url, the template can be removed?
```

Reply:
`url_template` can just be a URL or a template. See the [Azure documentation](https://docs.microsoft.com/en-us/azure/azure-signalr/concept-upstream),

> 
> These predefined parameters can be used in the URL pattern. Parameters will be replaced with a specified value when you're evaluating the upstream URL. 
> 
> For example: http://host.com/{hub}/api/{category}/{event}
> 
> When a client connection in the "chat" hub is connected, a message will be sent to this URL: http://host.com/chat/api/connections/connected

Or, if you provide category, event, or hub patterns those will be substituted in the respective places.

Reply:
that makes sense to me!